src/vcpkg/base/files.cpp

```diff
@@ -2250,9 +2207,146 @@ namespace vcpkg
                                CopyOptions options,
                                std::error_code& ec) override
         {
+#if defined(_WIN32)
             return stdfs::copy_file(
                 to_stdfs_path(source), to_stdfs_path(destination), convert_copy_options(options), ec);
+#else // ^^^ _WIN32 // !_WIN32 vvv
+            PosixFd source_fd{source.c_str(), O_RDONLY, ec};
+            if (ec)
+            {
+                return false;
+            }
+
+            struct stat source_stat;
+            source_fd.fstat(&source_stat, ec);
+            if (ec)
+            {
+                return false;
+            }
+
+            if (!S_ISREG(source_stat.st_mode))
+            {
+                // N4861 [fs.op.copy.file]/4: "report an error if"
+                // is_regular_file(from) is false
+                ec = std::make_error_code(std::errc::invalid_argument);
+                return false;
+            }
+
+            bool destination_exists;
+            struct stat destination_stat;
+            if (lstat(destination.c_str(), &destination_stat) == 0)
```

Comment:
I really really really really dislike this TOCTOU; can we not do something like:

```cxx
int options = O_WRONLY | O_CREAT;
if (masked_options == CopyOptions::overwrite_existing)
{
    options |= O_TRUNC;
}
else
{
    options |= O_EXCL;
}

PosixFd destination_fd{destination.c_str(), options, mode, ec};
if (ec)
{
    if(masked_options == CopyOptions::skip_existing && ec == std::errc::file_exists)
    {
        ec.clear();
    }
    return false;
}

struct stat destination_stat;
destination_fd.fstat(&destination_stat, ec);
if (ec)
{
    return false;
}
else if (!S_ISREG(destination_stat.st_mode)
{
    ec = std::make_error_code(std::errc::invalid_argument);
    return false;
}
```

Reply:
>can we not do something like:

* That does the wrong thing if destination is a symbolic link to a file, because `O_CREAT | O_EXCL` on a symbolic link, regardless of the target, fails: https://linux.die.net/man/3/open

Although looking now I'm incorrectly doing lstat...

* That does the wrong thing for `equivalent(source, dest)` && overwrite_existing: the standard behavior will do nothing but this will truncate source (!!!)

Reply:
After discussion out of band I think we fixed this.